flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/RowDataToProtoSerializer.java

```diff
@@ -0,0 +1,256 @@
+import org.apache.flink.table.types.logical.LocalZonedTimestampType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.TimeType;
+import org.apache.flink.table.types.logical.TimestampType;
+
+import com.google.api.client.util.Preconditions;
+import com.google.cloud.bigquery.storage.v1.BigDecimalByteStringEncoder;
+import com.google.cloud.flink.bigquery.sink.exceptions.BigQuerySerializationException;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.DynamicMessage;
+import org.apache.avro.generic.GenericRecord;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** Serializer for converting Flink's {@link RowData} to BigQuery proto. */
+public class RowDataToProtoSerializer extends BigQueryProtoSerializer<RowData> {
+
+    private Descriptor descriptor;
+    private LogicalType type;
+
+    public RowDataToProtoSerializer() {}
+
+    /**
+     * Prepares the serializer before its serialize method can be called. It allows contextual
+     * preprocessing after constructor and before serialize. The Sink will internally call this
+     * method when initializing itself.
+     *
+     * @param bigQuerySchemaProvider {@link BigQuerySchemaProvider} for the destination table.
+     */
+    @Override
+    public void init(BigQuerySchemaProvider bigQuerySchemaProvider) {
+        Preconditions.checkNotNull(
+                bigQuerySchemaProvider,
+                "BigQuerySchemaProvider not found while initializing AvroToProtoSerializer");
+        Descriptor derivedDescriptor = bigQuerySchemaProvider.getDescriptor();
+        Preconditions.checkNotNull(
+                derivedDescriptor, "Destination BigQuery table's Proto Schema could not be found.");
+        this.descriptor = derivedDescriptor;
+    }
+
+    public void setLogicalType(LogicalType type) {
+        this.type = type;
+    }
+
+    @Override
+    public ByteString serialize(RowData record) throws BigQuerySerializationException {
+        try {
+            return getDynamicMessageFromRowData(record, this.descriptor, this.type).toByteString();
+        } catch (Exception e) {
+            throw new BigQuerySerializationException(e.getMessage());
+        }
+    }
+
+    public Object toProtoValue(
+            LogicalType fieldType,
+            int fieldNumber,
+            RowData element,
+            FieldDescriptor fieldDescriptor) {
+        try {
+            // Logical Type contains the supposed type (since it is formed from BQ Table)
+            switch (fieldType.getTypeRoot()) {
+                case CHAR:
+                case VARCHAR:
+                    return element.getString(fieldNumber).toString();
```

Comment:
Why do you have to call a toString on a String?

Reply:
`element.getString()` returns `StringData` (https://nightlies.apache.org/flink/flink-docs-master/api/java/org/apache/flink/table/data/StringData.html) object, needs to be converted to string.

Reply:
Okay, you can close this comment.